Network/Converter/PacketConverter.cs

```diff
@@ -27,321 +28,680 @@
+                    .ToList<object>()
+                    .ForEach(primitiveElement =>
+                    {
+                        dynamic primitiveValue = primitiveElement;
+                        binaryWriter.Write(primitiveValue);
+                    });
             }
         }
 
+        #endregion Serialisation
+
+        #region Deserialisation
+
         /// <summary>
-        /// Reads the length of the array from the stream and creates the array instance.
-        /// Also fills the array by using recursion.  
+        /// Deserialises all the properties on the given <see cref="object"/> that
+        /// can be deserialised from the given <see cref="BinaryReader"/>s
+        /// underlying <see cref="MemoryStream"/>.
         /// </summary>
-        /// <param name="obj">The object.</param>
-        /// <param name="propertyInfo">The property information.</param>
-        /// <param name="binaryReader">The binary reader.</param>
-        /// <returns>System.Array.</returns>
-        private Array ReadArrayFromStream(object obj, PropertyInfo propertyInfo, BinaryReader binaryReader)
+        /// <param name="obj">
+        /// The <see cref="object"/> whose properties to deserialise using the given
+        /// <see cref="BinaryReader"/>.
+        /// </param>
+        /// <param name="binaryReader">
+        /// The <see cref="BinaryReader"/> from whose underlying <see cref="MemoryStream"/>
+        /// to deserialise the properties of the given <see cref="object"/>.
+        /// </param>
+        /// <returns>
+        /// The given <see cref="object"/> with all deserialisable properties
+        /// set.
+        /// </returns>
+        /// <remarks>
+        /// This method can only deserialise properties that lack the custom
+        /// <see cref="PacketIgnorePropertyAttribute"/>. Any other properties
+        /// will be left at their default values.
+        /// </remarks>
+        private object DeserialiseObjectFromReader(object obj, BinaryReader binaryReader)
         {
-            int arraySize = binaryReader.ReadInt32();
-            Type arrayType = propertyInfo.PropertyType.GetElementType();
-            Array propertyData = Array.CreateInstance(arrayType, arraySize);
-            for (int i = 0; i < arraySize; i++)
+            PropertyInfo[] propertiesToSerialise = GetTypeProperties(obj.GetType());
+
+            for (int i = 0; i < propertiesToSerialise.Length; ++i)
             {
-                if (arrayType.IsClass && !IsPrimitive(arrayType)) propertyData.SetValue(ReadObjectFromStream(Activator.CreateInstance(arrayType), binaryReader), i);
-                else propertyData.SetValue(ReadPrimitiveFromStream(arrayType, binaryReader), i);
+                propertiesToSerialise[i].SetValue(
+                    obj,
+                    DeserialiseObjectFromReader(
+                        obj, propertiesToSerialise[i], binaryReader));
             }
 
-            return propertyData;
+            return obj;
```

Comment:
obj could mean anything. Use meaningful variable names.

Reply:
Aha! This is where documentation comes in useful (:D) ! Since 'obj' is a transient value, and only used in the deserialisation chain it doesnt need to have a meaningful name. It represents the temprary object that is being deserialised, and since it is a parameter, upon hovering over it you will get the documentation. Also, i think that the name *is* appropriate, as it represents the object in the process of being deserialised - we cannot use 'object' as it is a keyword and thus reserved. Am willing to change it though.